File: trab_lua_lp/Catalogo.lua
-- CLASSE CATÁLOGO
-- CLASSE RESPONSÁVEL POR CONTROLAR TODAS AS FUNÇÕES E ATRIBUTOS DE CATÁLOGO
-- @AUTHOR IGOR VENTORIM
-- DATE BEGIN: 04/30/2016

dofile("Livro.lua")

local catalogoMethods = 
{
	-- # FUNÇÃO QUE ADICIONA UM LIVRO AO CATÁLOGO
	addLivro = function(self,cod,titulo,autor,assunto,data,editora,resumo)
		local livro = Livro(cod,titulo,autor,assunto,data,editora,resumo)
		self.livros[tonumber(cod)] = livro
	end,

	-- #FUNÇÃO QUE VERIFICA SE UM LIVRO JÁ PERTENCE AO CATÁLOGO
	isElementTable = function(self,cod)
		if(self.livros[cod]) then
			return 1
		else
			return 0
		end
	end,

	-- # FUNÇÃO QUE REMOVE UM LIVRO DA LISTA
	removeLivro = function(self,cod)
		self.livros[cod] = nil
	end,

	-- #FUNÇÃO QUE RETORNA O CATÁLOGO ATUAL
	getCatalogo = function(self)
		aux = {}
		count = 0
		for k, v in pairs(self.livros) do
			aux[count] = v
			count = count + 1
		end
		return aux
	end,

	-- # FUNÇÃO AUXILIAR PARA PRINTAR O CÓDIGO DE TODOS OS ITENS DO CATÁLOGO
	printCatalogo = function(self)
		for k, v in pairs(self.livros) do
			v:Print()
		end
	end,

	-- # FUNÇÃO QUE INFORMA O TAMANHO DO CATALOGO
	sizeCatalogo = function(self)
		count = 0
		for k, v in pairs(self.livros) do
			count = count + 1
		end
		return count
	end,

	-- #FUNÇÃO QUE ORDENA A LISTA CONFORME A FUNÇÃO DE ORDENAÇÃO DESEJADA
	ordenaCatalogo = function(self,cmpFunc)
		aux = self:getCatalogo()
		for j=1,#aux do
			chave = aux[j]
			i = j - 1
			while (i >= 0) and cmpFunc(aux[i],chave) == 1 do
				aux[i+1] = aux[i]
				i = i - 1
			end 
			aux[i+1] = chave
		end
		return aux	
	end
}


local catalogoMetatable =
{
	__index = catalogoMethods
}

-- #MÉTODO CONSTRUTOR DE CATÁLOGO
Catalogo = function()
	self = {}
	self.livros = {}
	setmetatable(self,catalogoMetatable)
	return self
end

File: trab_python_lp/Catalogo.py
#!/usr/bin/env python
# -'- coding: utf-8 -'-
from Livro import Livro

class Catalogo:

	#MÉTODO CONSTRUTOR DE CATÁLOGO
	def __init__(self):
		self.__livros = []

	# FUNÇÃO QUE ADICIONA UM LIVRO AO CATÁLOGO
	def addLivro(self,cod,titulo,autor,assunto,data,editora,resumo):
		livro = Livro(cod,titulo,autor,assunto,data,editora,resumo)
		self.__livros.append(livro)

	#FUNÇÃO QUE VERIFICA SE UM LIVRO JÁ PERTENCE AO CATÁLOGO
	def verificaLivro(self,code):
		for i in self.__livros:
			if(i.getCod() == code):
				return 1
		return 0

	
	# FUNÇÃO QUE REMOVE UM LIVRO DA LISTA
	# * CODE - Código identificador do livro
	def removeLivro(self,code):
		for i in self.__livros:
			if(i.getCod() == code):
				self.__livros.remove(i)
				return 1
		return 0

	# FUNÇÃO AUXILIAR PARA PRINTAR O CÓDIGO DE TODOS OS ITENS DO CATÁLOGO
	def printCatalogo(self):
		for i in self.__livros:
			print (i.getCod())

	#FUNÇÃO QUE RETORNA O CATÁLOGO ATUAL
	def getCatalogo(self):
		return self.__livros 

	#FUNÇÃO QUE ORDENA A LISTA CONFORME A FUNÇÃO DE ORDENAÇÃO DESEJADA
	def ordenaLista(self,cmpFunc):
	    for j in range(1, len(self.__livros)):
	        chave = self.__livros[j]
	        i = j - 1
	        while i >= 0 and cmpFunc(self.__livros[i],chave):
	            self.__livros[i+1] = self.__livros[i]
	            i -= 1
	        self.__livros[i + 1] = chave
	    return self.__livros